vendor/bat-native-ledger/src/bat/ledger/internal/publisher/server_publisher_fetcher.cc

```diff
@@ -0,0 +1,317 @@
+  }
+
+  if (!banner_details.logo_url().empty()) {
+    banner->logo = "chrome://rewards-image/" + banner_details.logo_url();
+  }
+
+  for (auto& amount : banner_details.donation_amounts()) {
+    banner->amounts.push_back(amount);
+  }
+
+  if (banner_details.has_social_links()) {
+    auto& links = banner_details.social_links();
+    if (!links.youtube().empty()) {
+      banner->links.insert(std::make_pair("youtube", links.youtube()));
+    }
+    if (!links.twitter().empty()) {
+      banner->links.insert(std::make_pair("twitter", links.twitter()));
+    }
+    if (!links.twitch().empty()) {
+      banner->links.insert(std::make_pair("twitch", links.twitch()));
+    }
+  }
+
+  return banner;
+}
+
+ledger::ServerPublisherInfoPtr ServerPublisherInfoFromMessage(
+    const publishers_pb::ChannelResponseList& message,
+    const std::string& expected_key) {
+  for (auto& entry : message.channel_responses()) {
+    if (entry.channel_identifier() != expected_key) {
+      continue;
+    }
+
+    auto server_info = ledger::ServerPublisherInfo::New();
+    server_info->publisher_key = entry.channel_identifier();
+    server_info->status = PublisherStatusFromMessage(entry);
+    server_info->address = entry.wallet_address();
+    server_info->updated_at =
+        static_cast<uint64_t>(base::Time::Now().ToDoubleT());
+
+    if (entry.has_site_banner_details()) {
+      server_info->banner =
+          PublisherBannerFromMessage(entry.site_banner_details());
+    }
+
+    return server_info;
+  }
+
+  return nullptr;
+}
+
+bool RemovePadding(base::StringPiece* padded_string) {
+  if (!padded_string) {
+    return false;
+  }
+
+  // Read payload length from the header.
+  if (padded_string->size() < sizeof(uint32_t)) {
+    return false;
+  }
+  uint32_t data_length;
+  base::ReadBigEndian(padded_string->data(), &data_length);
+
+  // Remove length header.
+  padded_string->remove_prefix(sizeof(uint32_t));
+  if (padded_string->size() < data_length) {
+    return false;
+  }
+
+  // Remove padding.
+  padded_string->remove_suffix(padded_string->size() - data_length);
+  return true;
+}
+
+bool DecompressMessage(base::StringPiece payload, std::string* output) {
+  constexpr size_t buffer_size = 32 * 1024;
+  return braveledger_helpers::DecodeBrotliStringWithBuffer(
+      payload,
+      buffer_size,
+      output);
+}
+
+}  // namespace
+
+namespace braveledger_publisher {
+
+ServerPublisherFetcher::ServerPublisherFetcher(
+    bat_ledger::LedgerImpl* ledger)
+    : ledger_(ledger) {
+  DCHECK(ledger);
+}
+
+ServerPublisherFetcher::~ServerPublisherFetcher() = default;
+
+void ServerPublisherFetcher::Fetch(
+    const std::string& publisher_key,
+    ledger::GetServerPublisherInfoCallback callback) {
+  CallbackVector& callbacks = callback_map_[publisher_key];
+  callbacks.push_back(callback);
+  if (callbacks.size() > 1) {
+    BLOG(1, "Fetch already in progress for publisher " << publisher_key);
```

Comment:
The `Fetch` method can be called from various places that want updated publisher data, so it's important to dedupe any requests that come in for the same publisher key.